src/kaleidoscope/device/dygma/Raise.cpp

```diff
@@ -0,0 +1,599 @@
+  if (!key_addr.isValid())
+    return false;
+
+  auto row = key_addr.row();
+  auto col = key_addr.col();
+
+  if (col >= 8) {
+    return rightHandMask.rows[row] & (1 << (7 - (col - 8)));
+  } else {
+    return leftHandMask.rows[row] & (1 << (7 - col));
+  }
+}
+
+void RaiseKeyScanner::maskHeldKeys() {
+  memcpy(leftHandMask.rows, leftHandState.rows, sizeof(leftHandMask));
+  memcpy(rightHandMask.rows, rightHandState.rows, sizeof(rightHandMask));
+}
+
+bool RaiseKeyScanner::isKeyswitchPressed(KeyAddr key_addr) {
+  auto row = key_addr.row();
+  auto col = key_addr.col();
+
+  if (col >= 8) {
+    return (bitRead(rightHandState.rows[row], 15 - col) != 0);
+  } else {
+    return (bitRead(leftHandState.rows[row], col) != 0);
+  }
+}
+
+bool RaiseKeyScanner::wasKeyswitchPressed(KeyAddr key_addr) {
+  auto row = key_addr.row();
+  auto col = key_addr.col();
+
+  if (col >= 8) {
+    return (bitRead(previousRightHandState.rows[row], 15 - col) != 0);
+  } else {
+    return (bitRead(previousLeftHandState.rows[row], col) != 0);
+  }
+}
+
+uint8_t RaiseKeyScanner::pressedKeyswitchCount() {
+  return __builtin_popcountl(leftHandState.all) + __builtin_popcountl(rightHandState.all);
+}
+
+uint8_t RaiseKeyScanner::previousPressedKeyswitchCount() {
+  return __builtin_popcountl(previousLeftHandState.all) + __builtin_popcountl(previousRightHandState.all);
+}
+
+void RaiseKeyScanner::setKeyscanInterval(uint8_t interval) {
+  RaiseHands::leftHand.setKeyscanInterval(interval);
+  RaiseHands::rightHand.setKeyscanInterval(interval);
+}
+
+void RaiseKeyScanner::setup() {
+  // first set all pins to outputs and low - EMC considerations
+  // not serial pins
+  for (int pin = 2; pin < 27; pin ++) {
+    pinMode(pin, OUTPUT);
+    digitalWrite(pin, LOW);
+  }
+  // not USB pins
+  for (int pin = 30; pin < 43; pin ++) {
+    pinMode(pin, OUTPUT);
+    digitalWrite(pin, LOW);
+  }
+}
+
+void RaiseKeyScanner::reset() {
+  leftHandState.all = 0;
+  rightHandState.all = 0;
+  kaleidoscope::hid::releaseAllKeys();
+  kaleidoscope::hid::sendKeyboardReport();
+}
+
+/********* Hardware plugin *********/
+void Raise::setup() {
+  RaiseHands::setup();
+  KeyScanner::setup();
+  LEDDriver::setup();
+
+  // initialise Wire of scanner - have to do this here to avoid problem with
+  // static object intialisation ordering
+  twi_init(I2C_CLOCK_KHZ);
+
+  RaiseHands::initializeSides();
+}
+
+void Raise::side::prepareForFlash() {
+  Wire.end();
+
+  power(LOW);
+  // also turn off i2c pins to stop attiny from getting enough current through i2c to stay on
+  pinMode(SCL, OUTPUT);
+  pinMode(SDA, OUTPUT);
+  digitalWrite(SCL, false);
+  digitalWrite(SDA, false);
+
+  // wipe key states, to prevent accidental key repeats
+  RaiseKeyScanner::reset();
+
+  power(HIGH);
+
+  Wire.begin();
+  Wire.setClock(I2C_FLASH_CLOCK_KHZ * 1000);
+  // wait for side bootloader to be ready
+  delay(100);
+}
+
+void Raise::side::reinitAfterFlash(uint8_t addr) {
```

Comment:
I think this is more like just "reinit" or "boot" or "bootFromBootloader" ...but should also just magically happen on the ATTiny side.

Reply:
I haven't been able to _test_ this, as my Raise PCB is too old (I think... flashing the attinys doesn't work even with the factory firmware). As soon as I can test, I'll verify. For the time being, I guess we can drop this, and I'll re-add it if it turns out to be required.